### PR TITLE
Add witness service always

### LINKED ISSUE #320: Bug: failed partial validation with error: 'bad-witness-nonce-size : invalid witness nonce size'.
When Segwit has been enabled on the network through BIP9 flags, neither the peer or the node will announce it supports Segwit as a feature. This will force the node to strip Segwit data and send that to the peer, even though the peer supports Segwit. This will cause the peer to log the following errors:

warn: Blockcore.Features.Consensus.Rules.PosConsensusRuleEngine[0]
      Block '46-bd16c243b1a9e67baeb9bca6da07d6a9bc0206109e8f8b9e50e7882338926672-HeaderValidated - PH' failed partial validation with error: 'bad-witness-nonce-size : invalid witness nonce size'.

warn: Blockcore.Consensus.ConsensusManager[0]
      Downloading block for '46-bd16c243b1a9e67baeb9bca6da07d6a9bc0206109e8f8b9e50e7882338926672-HeaderValidated - PH' had missing services NODE_WITNESS, it will be enqueued again.

### PR DESCRIPTION
**Pull request by sondreb**
- Add the witness service on startup to the node.
- Refactor the "WithOptions" method to clearly state conditions for logic.
- Closes #320

### CODE REVIEW

File: src/Features/Blockcore.Features.Consensus/ConsensusFeature.cs [hunk 1]
```diff
@@ -56,11 +56,11 @@ public override Task InitializeAsync()
             {
                 // Add witness discovery as a requirement if witness is activated.
                 this.connectionManager.AddDiscoveredNodesRequirement(NetworkPeerServices.NODE_WITNESS);
-
-                // Set witness as a supported service if witness is activated.
-                this.connectionManager.Parameters.Services |= NetworkPeerServices.NODE_WITNESS;
             }
 
+            // Always announce that the node supports WITNESS even if witness is not activated yet.
+            this.connectionManager.Parameters.Services |= NetworkPeerServices.NODE_WITNESS;
```
Comment (dangershony) [inline src/Features/Blockcore.Features.Consensus/ConsensusFeature.cs:62]: I think its correct to only put that on service , it means when segwit is not active the peer will still connect to older nodes that support segwit but dont broadcast that

**Comment by dangershony** [comment]
Should we delete the class `HaveWitness`?

**Comment by sondreb** [comment]
> Should we delete the class `HaveWitness` as it never used?

Yes, I think so, will avoid any confusion about it.

**Comment by MithrilMan** [comment]
What's wrong imo about the original code is that it loads NODE_WITNESS based on the node current consensus tip, so if the witness gets activated while the node is running, the NODE_WITNESS flag isn't set until the node is restarted

maybe that required service will get set by this code but I'm not sure
https://github.com/block-core/blockcore/blob/0695789f1e4729ca9beab137b40b9c81a3e8dce9/src/Blockcore/Consensus/ConsensusManager.cs#L460-L461

same thing for the in the same InitializeAsync method
```
// Set witness as a supported service if witness is activated.
this.connectionManager.Parameters.Services |= NetworkPeerServices.NODE_WITNESS;`
```

beside this (that would happen only when a deployment activates (but it's still a bug imo), I don't think you have to touch Block.cs at all but just set properly the NODE_WITNESS flag if your node version supports it (no matter if it has been activated or not)

When you set a service in your supported node services you are saying to other nodes that your node support that features, it doesn't mean you don't have to check activation within the code

Can you try to do a test?
try to run a network with an activation at a specified height, then when the height has been reached (and you likely hit the exception) restart the node, it should work (as long as your consensus tip is higher than the activation point)

Anyway I think that the change to move the `this.connectionManager.Parameters.Services |= NetworkPeerServices.NODE_WITNESS;` out of the if is correct, (probably it should even check protocol version but I don't see it as a problem if you signal a service that the other node wouldn't understand anyway)

**Comment by sondreb** [comment]
Restarting the node will make it announce segwit support, but unless restarted it will strip out segwit data when blocks are sent to the connected peers, causing the errors logges in #320 

Changes to the Block.cs was made to make it more clear what condition is triggered to make it strip out segwit data, it did not change any logic.

**Comment by dangershony** [comment]
> Can you try to do a test?
try to run a network with an activation at a specified height, then when the height has been reached (and you likely hit the exception) restart the node, it should work (as long as your consensus tip is higher than the activation point)  

This we did and restart does indeed work because then the node will signal the witness services in the handshake version.
This bug is very specific to when segwit activates and nodes are currently connected.

 >  if (validationContext.MissingServices != null) 
     this.connectionManager.AddDiscoveredNodesRequirement(validationContext.MissingServices.Value); 

the above wont help (we tried too) in theory it should disconnect the node and then reconnection will signal segwit services but its not so straight forward and besides we also agree that specifying segwit in services on startup is the correct way and it does fix this issue.

**Comment by MithrilMan** [comment]
> Restarting the node will make it announce segwit support, but unless restarted it will strip out segwit data when blocks are sent to the connected peers, causing the errors logges in #320
> 
> Changes to the Block.cs was made to make it more clear what condition is triggered to make it strip out segwit data, it did not change any logic.

you are just confirming what I said then
also if you check bitcoin core source you can find this

https://github.com/bitcoin/bitcoin/blob/01bb3afb5197cffec6bfe324a2efc314ddfdca57/src/init.cpp#L1836-L1840
```
    if (chainparams.GetConsensus().SegwitHeight != std::numeric_limits<int>::max()) {
        // Advertise witness capabilities.
        // The option to not set NODE_WITNESS is only used in the tests and should be removed.
        nLocalServices = ServiceFlags(nLocalServices | NODE_WITNESS);
    }
```

please remove Blocks.cs changes because it change current code, you removed
```
if ((options == TransactionOptions.Witness) && this.Transactions[0].HasWitness)
                return this;
```

and this saves you to rebuild the block if you have already what you need to send

just announcing always your support for NODE_WITNESS should be enough

**Comment by sondreb** [comment]
The code was changed from having multiple guard-clauses that required some thinking (reading), to having a single if condition (and one guard) and then does "return this" on else, maybe it's better to view the source after it was changed and not the diff.

```
if ((options == TransactionOptions.Witness) && this.Transactions[0].HasWitness)
                return this;
```

This is not needed, the conditions goes to "else".

**Comment by MithrilMan** [comment]
> The code was changed from having multiple guard-clauses that required some thinking (reading), to having a single if condition (and one guard) and then does "return this" on else, maybe it's better to view the source after it was changed and not the diff.
> 
> ```
> if ((options == TransactionOptions.Witness) && this.Transactions[0].HasWitness)
>                 return this;
> ```
> 
> This is not needed, the conditions goes to "else".

yeah missed the else and that's exactly the reason why we have always (almost) guards on top of code, because the flow is easier to follow than having an else at the end.

You can easily solve this issue just by early returning by negating your `if`, you can then remove the else and the problem is solved

Anyway my point is that that code doesn't need a change to fix the issue

**Comment by dangershony** [comment]
> please remove Blocks.cs changes because it change current code, you removed

no we didn't remove it we just added a condition and realized the inverse is more efficient.
Basically we assume that the only reason to strip witness data is when the peer is not supporting witness all other cases block should stay as it is (the exception is witness not activated and and block has witness data which is impossible)

**Comment by dangershony** [comment]
> With your change if you have to serialize with witness but the transaction doesn't have witness, you are using an incorrect block.

But how? that's not possible, I think we can assume if witness is active and our node supports witness then it much have witness data the block.

**Comment by sondreb** [comment]
> Are you sure about that change?
> `if ((options == TransactionOptions.Witness) && this.Transactions[0].HasWitness)` ensures that if your option is to send witness and your transaction already have witness, you don't need to regenerate the block with withness information.
> 
> With your change if you have to serialize with witness but the transaction doesn't have witness, you are using an incorrect block.

The logic is opposite of what you think, it re-generate the transaction WITHOUT segwit data if the peer does not support it. If option to enable witness is on, we'll simply return what is made and make no changes, no matter if segwit data is there or not.

**Comment by MithrilMan** [comment]
To me the original code says

- if you want witness and block has witness, fine, return this
- if you don't want witness and block does not have witness, fine, return this

otherwise rebuild the block with what you want
```
 bms = new BitcoinStream(ms, false, consensusFactory)
                {
                    TransactionOptions = options,
                };
                ```
that's different that what you are saying

**Comment by dangershony** [comment]
> If option to enable witness is on, we'll simply return what is made and make no changes, no matter if segwit data is there or not

yes indeed and thats because we now changed the behavior slightly of the flag, before the witness_services was on when segwit is activated now its on all the time, and make the assumption we always have blocks corresponding to what we support (i.e we have witness we have block with witness data)

so the only reason to reserailize he block is when we need to strip off witness data we will never add to it witness data (beides I don't think the serializer can do that, it can't add witness data this has to come from the originally constructed witness block)

**Comment by dangershony** [comment]
I'd prefer to revert back the one `if` condition we had once @MithrilMan is convinced.

**Comment by MithrilMan** [comment]
> I'd prefer to revert back the one if condition we had once @MithrilMan is convinced.

I'm not against a single if, I'm against having a final else with a `return this`
it can be easily solved by just negating the single if and returning `this` there, so it's a guard and you don't need to wrap the other code in brackets nor having an else statement

**Review by dangershony — APPROVED** [review]

**Comment by dangershony** [comment]
I suppose @MithrilMan is now in agreement then.
And seems tests passed too but perhaps its a good idea to also run tests locally on this PR so I will do it now (hold on the commit)

**Comment by dangershony** [comment]
All tests passed locally

**Pull request merged**